src/main/java/ch/njol/skript/expressions/ExprSignText.java

```diff
@@ -18,159 +18,298 @@
+		boolean acceptsMany = line == null;
+		switch (mode) {
+			case REMOVE:
+			case REMOVE_ALL:
+			case DELETE:
+				acceptsMany = false;
+			case SET:
+			case ADD:
+				return CollectionUtils.array(acceptsMany ? String[].class : String.class);
+			case RESET:
+			default:
+				return null;
+		}
 	}
-	
-	static boolean hasUpdateBooleanBoolean = true;
-	
-	@SuppressWarnings("incomplete-switch")
+
 	@Override
-	public void change(final Event e, final @Nullable Object[] delta, final ChangeMode mode) throws UnsupportedOperationException {
-		final Number l = line.getSingle(e);
-		if (l == null)
-			return;
-		final int line = l.intValue() - 1;
-		if (line < 0 || line > 3)
-			return;
-		final Block b = block.getSingle(e);
-		if (b == null)
+	public void change(Event event, @Nullable Object[] delta, ChangeMode mode) {
+		int line = 0;
+		if (this.line == null && !lines) {
 			return;
-		if (getTime() >= 0 && e instanceof SignChangeEvent && b.equals(((SignChangeEvent) e).getBlock()) && !Delay.isDelayed(e)) {
-			switch (mode) {
-				case DELETE:
-					((SignChangeEvent) e).setLine(line, "");
-					break;
-				case SET:
-					assert delta != null;
-					((SignChangeEvent) e).setLine(line, (String) delta[0]);
-					break;
-			}
-		} else {
-			if (!sign.isOfType(b))
+		} else if (!lines) {
+			line = this.line.getOptionalSingle(event).orElse(1).intValue() - 1;
+			if (line < 0 || line > 3)
 				return;
-			final Sign s = (Sign) b.getState();
+		}
+		if (getTime() >= 0 && event instanceof SignChangeEvent && blocks.check(event, block -> block.equals(((SignChangeEvent) event).getBlock())) && !Delay.isDelayed(event)) {
+			String[] stringDelta = delta == null ? null : Arrays.copyOf(delta, delta.length, String[].class);
+			SignChangeEvent changeEvent = (SignChangeEvent) event;
 			switch (mode) {
+				case ADD:
+					assert stringDelta != null;
+					if (lines) {
+						List<String> list = Lists.newArrayList(changeEvent.getLines());
+						for (String string : stringDelta)
+							list.add(string);
+
+						stringDelta = list.toArray(new String[0]);
+					} else {
+						for (int i = 0; i < stringDelta.length; i++) {
+							String value = stringDelta.length > i ? (String) delta[i] : "";
+							stringDelta[i] = value + stringDelta[i];
+						}
+					}
+					//$FALL-THROUGH$
+				case REMOVE:
+				case REMOVE_ALL:
+					assert stringDelta != null;
+					stringDelta = ExprLore.handleRemove(StringUtils.join(stringDelta, "\n"), stringDelta[0], mode == ChangeMode.REMOVE_ALL).split("\n");
+					//$FALL-THROUGH$
 				case DELETE:
-					s.setLine(line, "");
-					break;
+					stringDelta = CollectionUtils.array("", "", "", "");
 				case SET:
-					assert delta != null;
-					s.setLine(line, (String) delta[0]);
+					// We need to ensure that it's clearing values without calling setLine twice.
+					if (mode == ChangeMode.SET) {
+						for (int i = 0; i < 4; i++)
+							stringDelta[i] = stringDelta.length > i ? (String) stringDelta[i] : "";
+					}
```

Comment:
Do we really need to do this if we're only setting one line?
(we do it anyway a few lines down if we're setting multiple lines)